test/integration/graphql_test.lua

```diff
@@ -687,6 +736,174 @@ g.test_custom_type_scalar_variables = function()
     end)
 end
 
+g.test_output_type_mismatch_error = function()
+    local server = cluster.main_server
+
+    server.net_box:eval([[
+        local fiber = require('fiber')
+        package.loaded['test'] = {}
+        package.loaded['test']['expected_list'] = function(_, args)
+            if args.type == 'boolean' then
+                return {{value = {true}}}
+            elseif args.type == 'boolean_list' then
+                return {{value = {{value = true}}}}
+            elseif args.type == 'string' then
+                return {{value = {'Hello'}}}
+            elseif args.type == 'number' then
+                return 123
+            elseif args.type == 'cdata' then
+                return 123LL
+            elseif args.type == 'userdata' then
+                return fiber.create(function() end)
+            elseif args.type == 'object' then
+                return {a = 'b'}
+            end
+            return {{value = {{value = {1}}}}, {value = {{value = {2}}}}}
+        end
+
+        local graphql = require('cartridge.graphql')
+        local types = require('cartridge.graphql.types')
+
+        local nested_type = types.object({
+            name = 'NestedObjectWithValue',
+            fields = {
+                value = types.list(types.int),
+            },
+        })
+
+        local obj_type = types.object({
+            name = 'ObjectWithValue',
+            fields = {
+                value = types.list(nested_type),
+            },
+        })
+
+        graphql.add_callback({
+            name = 'expected_list',
+            args = {
+                type = types.string,
+            },
+            kind = types.list(obj_type),
+            callback = 'test.expected_list',
+        })
+    ]])
+
+    t.assert_equals(
+        server:graphql({
+            query = [[
+                query($type: String) {
+                    expected_list(type: $type) { value { value } }
+                }
+            ]],
+        variables = {type = box.NULL}}
+        ).data.expected_list, {{value = {{value = {1}}}}, {value = {{value = {2}}}}}
+    )
+
+    t.assert_error_msg_equals('Expected a table for field "value"', function()
+        return server:graphql({
+            query = [[
+                query($type: String) {
+                    expected_list(type: $type) { value { value } }
+                }
+            ]],
+        variables = {type = 'boolean'}}
```

Comment:
You are testing error messages here. Declaring variables looks unnecessary too. Same for the following tests

Reply:
Variable value changes return value. And I see no bad to use graphql variables in such case.

Reply:
It's not bad, it's just not necessary. You can put this value inside `type` argument without using variables.

Reply:
Removed